tests/test_project.py

```diff
@@ -2517,6 +2531,224 @@ def test_raise_exception_in_hook_cmd(self):
+        track.install_project_hooks(project)
+        assert len(project.project_hooks.on_start) == 1
+        assert len(project.project_hooks.on_success) == 1
+        assert len(project.project_hooks.on_exception) == 1
+
+
+class TestHooksTrackOperationsNotStrict(TestHooksSetUp):
+    project_class = define_hooks_track_operations_project._HooksTrackOperations
+    entrypoint = dict(
+        path=os.path.realpath(
+            os.path.join(
+                os.path.dirname(__file__), "define_hooks_track_operations_project.py"
+            )
+        )
+    )
+    log_fname = define_hooks_track_operations_project.LOG_FILENAME
+    error_on_no_git = False
+
+    @pytest.fixture(
+        params=[
+            ("base", define_hooks_track_operations_project.HOOKS_ERROR_MESSAGE),
+            ("cmd", "non-zero exit status 42"),
+        ],
+        ids=lambda x: x[0],
+    )
+    def operation_info(self, request):
+        return request.param
+
+    @pytest.fixture(
+        params=None if skip_git else (True, False),
+        ids=None if skip_git else ("git-dirty", "git-clean"),
+    )
+    def git_repo(self, project, request):
+        # params=None is equivalent to not passing a parameter which results in
+        # result.param being unset.
+        if not hasattr(request, "param"):
+            return
+        repo = git.Repo.init(project.path)
+        with open(project.fn("test.txt"), "w"):
+            pass
+        repo.index.add(["test.txt"])
+        repo.index.commit("Initial commit")
+        if request.param:
+            with open(project.fn("dirty.txt"), "w"):
+                pass
+            repo.index.add(["dirty.txt"])
+        return repo
+
+    def split_log(self, job, fn):
+        with open(job.fn(fn)) as f:
+            values = f.read().split("\n")
+        if values[-1] == "":
+            values = values[:-1]
+        return values
+
+    def get_log(self, job, fn=None):
+        if fn is None:
+            execution_history = job.doc["execution_history"]()
+            # Add back in redundent information to make check_metadata universal
+            for key, item in execution_history.items():
+                for entry in item:
+                    entry["job-operation"] = {"name": key, "job_id": job.id}
+            return execution_history
+        log_entries = [json.loads(entry) for entry in self.split_log(job, fn)]
+        execution_history = {}
+        for entry in log_entries:
+            operation_name = entry["job-operation"]["name"]
+            execution_history.setdefault(operation_name, []).append(entry)
+        return execution_history
+
+    def check_metadata(
+        self,
+        metadata,
+        job,
+        operation_name,
+        error_message,
+        before_execution,
+        expected_stage,
+        repo,
+    ):
+        assert metadata["stage"] == expected_stage
+        # When running on MacOS CI a private/ is prepended to job.project.path
```

Comment:
This is rather surprising behavior. I would encourage finding a way to have deterministic tests.

Reply:
I thought about this a bit more. I would xfail or skip this test on macOS rather than introduce a lot of workarounds in the testing code.